packages/core/src/batiq.ts

```diff
@@ -0,0 +1,93 @@
+import * as lens from "./lens";
+import {
+  AppOperation,
+  AppSchema,
+  BaseBatiqCore,
+  CreateBatiqFn,
+  Middleware,
+} from "./types";
+
+export const createBatiq: CreateBatiqFn = (
+  json: AppSchema,
+  middlewares: Middleware<BaseBatiqCore, BaseBatiqCore>[] = []
+) => {
+  const listeners = new Set<(operations: AppOperation[]) => void>();
+  const batiq: BaseBatiqCore = {
+    operations: [],
+
+    dispatch: (operations) => {
+      batiq.operations = batiq.operations.concat(operations);
+
+      const performOperation = (operation: AppOperation) => {
+        switch (operation.type) {
+          case "set": {
+            json = lens
+              .fromPath(json, operation.path as any)
+              .set(json, operation.value as any);
+            break;
+          }
+
+          case "remove": {
+            json = lens.mod(
+              json,
+              operation.path.slice(0, -1) as any,
+              (value: any) => {
+                const lastKey = operation.path[operation.path.length - 1];
+                if (Array.isArray(value)) {
+                  return value.filter((_, i) => i !== lastKey);
+                }
+                const { [lastKey]: _, ...rest } = value;
+                return rest;
+              }
+            );
+            break;
+          }
+
+          case "insert": {
+            const path = operation.path.slice(0, -1);
+            const index = operation.path[operation.path.length - 1];
+            json = lens.mod(json, path as any, (value: any): any =>
+              Array.isArray(value) && typeof index === "number"
+                ? [
+                    ...value.slice(0, index),
+                    operation.value,
+                    ...value.slice(index),
+                  ]
+                : value
+            );
+            break;
+          }
+
+          case "move": {
+            const value = lens.fromPath(json, operation.from as any).get(json);
+            if (!value) return;
+            performOperation({ type: "remove", path: operation.from });
+            performOperation({
+              type: "insert",
+              path: operation.to,
+              value,
+            });
+            break;
+          }
+        }
+      };
+      operations.forEach(performOperation);
+
+      [...listeners].forEach((listener) => listener(operations));
+    },
+
+    export: () => json,
```

Comment:
I've been thinking on user perspective, I think it's more declarative to use `getSchema()` or `getState()` for the name